File: Demo/Demo/Collections/Extensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Demo.Collections
{
	static class Extensions
	{
		public static IEnumerable<T> Where<T>(this IEnumerable<T> self, Func<T, bool> condition)
		{
			var result = new List<T>();

			foreach (var elem in self)
			{
				if (condition(elem))
				{
					result.Add(elem);
				}
			}

			return result;
		}

		public static IEnumerable<R> Select<T, R>(this IEnumerable<T> self, Func<T, R> action)
		{
			var result = new List<R>();

			foreach (var elem in self)
			{
				var r = action(elem);
				result.Add(r);
			}

			return result;
		}

		public static int Count<T>(this IEnumerable<T> self)
		{
			var result = 0;

			foreach (var elem in self)
			{
				result++;
			}

			return result;
		}
	}
}

File: Demo/DemoVisualBasic/Collections/Extensions.vb

Imports System.Collections.Generic
Imports System.Linq
Imports System.Runtime.CompilerServices
Imports System.Text
Imports System.Threading.Tasks

Namespace Demo.Collections
	Module Extensions
		<Extension>
		Public Function Where(Of T)(self As IEnumerable(Of T), condition As Func(Of T, Boolean)) As IEnumerable(Of T)
			Dim result = New List(Of T)()

			For Each elem In self
				If condition(elem) Then
					result.Add(elem)
				End If
			Next

			Return result
		End Function

		<Extension>
		Public Function [Select](Of T, R)(self As IEnumerable(Of T), action As Func(Of T, R)) As IEnumerable(Of R)
			Dim result = New List(Of R)()

			For Each elem In self
				Dim re = action(elem)
				result.Add(re)
			Next

			Return result
		End Function

		<Extension>
		Public Function Count(Of T)(self As IEnumerable(Of T)) As Integer
			Dim result = 0

			For Each elem In self
				result += 1
			Next

			Return result
		End Function
	End Module
End Namespace
